### DISCUSSION TITLE
Koriki for RG35XX Beta 0.9

### DISCUSSION THREAD

**Discussion by acmeplus** [rg35xx-cfw/Koriki#24] [Announcements]
# Introduction
This is probably the last major public beta release of Koriki for the RG35XX before 1.0 is released. No major changes (other than bug fixes) are expected from this 0.9 to 1.0.  

Note that this is a beta release, so while the system is usable there may be issues that still need to be solved. Use the [Discussions Page](https://github.com/rg35xx-cfw/Koriki/discussions) to discuss any issues you may find, and create issues in the [Issues Page](https://github.com/rg35xx-cfw/Koriki/issues)

# Installation:
* Flash the koriki-rg35xx image to a SDCARD using balena etcher or a similar software
* Boot with the SDCARD. The first time the firmware will expand the second partition and copy some firmware files to that partition
* Important: this version now uses FAT32 for the second partition, so it should be accessible on Mac/PC without any dedicated software
* If you use a second SDCARD, the system will install those files in the last partition of the second card
* Note that the first time the system boots it will take a bit longer to expand the card and copy files over. Even when the menu is up, there will be a few seconds when the system will be a bit slow because it's adding the virtual memory image
* After the first boot the system will create some configuration folders as well as the roms and bios folders. Add your roms to the roms folder, and your bios to the roms folder.

# Update from previous installations:

* NOTE: If you installed v0.7 or higher, it's safe to use the update file. If you installed v0.6 or a previous version, you should install the full image instead since the partition tables have changed.
* Uncompress the file boot.tar.xz, and replace the contents of the first partition of your SDCARD image with this one (you should see files like uImage, ramdisk.img, boot folder, etc.)
* If you run into issues with content not starting or any other issue, rename your system folder (from the last partition of your main SDCARD or your second SDCARD), and start again.

# IMPORTANT
* If you use a second SDCARD that you have previously used with koriki for RG35XX, please rename your system folder, and start again.
* Some RG35XX boards have issues booting koriki. Download the archive ``bootfix_for_some_revisions.zip``, extract it, and copy the files ``uImage`` and ``kernel.dtb`` to the first Koriki SDCARD partition (overwriting the existing ones). Note that while that fix may solve boot issues, it may introduce some inconsistencies regarding controls like volume keys. 

# Charger behavior:
* The charging logic is slightly different from the one in the stock firmware (or Garlic). The system reimplements the charger logic that is performed via software and not via hardware in this system.
* Shutdown:
  * If the charger power cable is connected to the USB C port, once the system is shutdown it will boot into the charger mode. You will briefly see a battery symbol and the system will seem to be off but the charger led will be on (orange). To boot from this state, briefly press power key.
  * If the charger cable is not connected, the system will be off completely. To boot again from this state you need to press the power button a few seconds and the system will boot.
* Reboot:
  * The system will reboot as expected
 
# ChangeLog:
2023/07/04:
* Fixed controls getting stuck in one direction. This was mostly visible on shoot’em up games like 1944, Last Resort, etc.
* Removed evmapy dependency. Added joystick support to simplemenu so evmapy is no longer needed to create key events from joystick events
* Retroarch no longer has duplicated or incorrect events due to the removal of evmapy and new input mappings. 
* Retroarch key mappings/hotkeys have changed to be consistent with Koriki on the Miyoo Mini:
    * MENU + A: Pause
    * MENU + B: FastForward Toggle
    * MENU + X: Menu
    * MENU + Y: Reset Content
    * MENU + L1: Save Save State
    * MENU + R1: Load Save State
    * MENU + L2: Next Save State
    * MENU + R2: Previous Save State
    * START + SELECT: Quit Game
* Created SDL_GAMECONTROLLERCONFIG configuration file in ~/.config/gamecontrolerconfigdb.txt
* Fixed the simplemenu framebuffer glitch when launching roms/ports
* Fixed simplemenu background changing when selecting cores in simplemenu (pressing select to access game menu)
* Fixed simplemenu  favorites not being created
* Fixed first SDCard SHARE partition not getting properly expanded (not visible on MacOS, etc.)
* Added RNDIS support (thanks @JohnnyOnFlame). You need to use the adb_uImage and then create a script that calls usbmond.sh RNDIS.
* Initial wifi work (not yet working via configuration files). 
* Bumped version to Beta 0.9

2023/06/19
* Added timeout for ``POWER`` button press. System no longer shutdowns immediately when the button is pressed, now it needs to be pressed about 3s
* Fixed volume increasing to 100% when unit was shut down via power button
* Fixed brightness control combo (``MENU``+``VOL+/-``)
* Fixed audio volume keys if you use adb_uImage (mixer was different between adb_uImage and uImage)
* Fixed HDMI audio. Note that HDMI cable needs to be plugged in before boot for HDMI to work
* Added PCSX standalone configuration
* Added Devolutionx (Diablo) gamepad support
* Re-Added KM MAME 2003 xtreme core (removed by mistake in 0.7.3)
* Added libretro CROCODS core (Amstrad CPC)
* Added libretro quickness core (NES)
* ~~Added libretro bk (elektronika PDP)~~ (currently broken)
* Added libretro gearcoleco (ColecoVision)
* Added libretro GME (blargg's Game_Music_Emu)
* Added libretro RACE (NGPC)
* Added launcher for eduke32 (DukeNukem3D)
* Bumped version to Beta 0.8

2023/06/16
* Modified the FP flags from VFP3-D16 to NEON due to some performance issues detected in PCSX rearmed core. This should bring some performance improvements
* Recompiled all cores with the new architecture
* Updated full buildroot packages to synchronize to the latest batocera build
* Removed some unnecessary patches. Note that there may be some inconsistencies with previous releases, please report those in the discussions page, discord, or create issues if the problem is well defined
* Improved charging logic. When connecting a power cable when the console is off, the console will enter into charging mode without booting into koriki
* Added PCSX standalone (note, currently is not possible to exit the emulator, if anyone knows how to quit, please let me know)
* Added default mupen64 controls (you can exit the emulator with ``START``+``SELECT``)
* Added ATC theme in Spanish (Thanks to @albgr1979) (theme switching should now work)
* Enabled USB for controllers and some other devices. By default this release works with USB devices like gamepads, USB Wifi and USB Ethernet. However, please not that USB wireless is not currently configured.
  * Note: If you want to use ADB, copy over ``adb_uImage`` into  ``uImage`` 
* Improved input handling by updating the input driver to the latest garlic gpio module (quick movement from L to R or U to D should not leave the key stuck in the initial direction)
  * Note: there's a regression with the power button due to the new input driver. Use the simplemenu menu to shutdown/reboot

2023/06/13
* Added power button shutdown. Short press power button will shutdown the console. If the charger is connected it will boot into charger mode. From there another short press boots the console
* Fixed shutdown while charging
* Added brightness control shortcut with ``MENU``+``VOLUME+/VOLUME-``
* Fixed Sega32X launcher
* Changed AmstradCPC core to cap32

2023/06/12
* Update to fix an issue with the swap and share partitions being overwritten
* Reapplied the SDL patch that was not correctly applied (thanks again @johnnyonflame)
* Fixed typo in mastersystem path
* Added proper openbor launchers
* Added o2em launcher
* Fixed sega32x path

2023/06/11
* Added SNES mednafen-supafaust core (default for SNES). StarFox now runs close to 60fps without overclock or video threading
* Added MAME km_mame_2003_xtreme that runs much better for some roms (default for MAME)
* Fixed swap virtual memory generation, it has its own dedicated partition now, should be much more efficient and not impact FAT32 partitions
* Added additional ports location
* Added SDL VT patch to solve some issues with some ports not returning to simplemenu (thanks @johnnyonflame)

2023/06/08
* Fixed simplemenu NO GAMES FOUND message. Now it will go back to the last selection after a game is closed or after restart
* Changed data partition to FAT32 format
* Added controller configuration for standalone emulators (flycast, PPSSPP, mupen64plus, and Drastic)
* Changed default emulators, removed some missing emulators from the core list

2023/06/06
* First public release (alpha v0.5)
* Implemented global audio control (thanks to [adixal](https://codeberg.org/adixal) for his muOS joyaudio driver)
* Improved the charger logic. System now should properly respond to shutdown or reboot commands in koriki/simplemenu. Just go to settings -> Session, press left/right until you see either shutdown or reboot and press A to confirm. See charger behavior below for details.
* Updated flycast and ppsspp cores to the latest release
* Added drastic, flycast, and ppsspp application launchers. 

2023/06/04
* Updated mgba core
* Added cores: vba-m, gpsp, stella2014
* Added flycast standalone
* Fixed partition autoexpand

2023/06/03
* Updated launchers and menu sections so most systems should now work right out of the box
* Fixed charger led indicators
* Added od-commander
* Added missing flycast_libretro cores
* Added virtual memory back (should help to launch N64 titles)

2023/06/02
* Updated ``koriki-rg35xx_20230602.img.gz``
* Fixed battery charge indicator

2023/06/01
* Reworked system settings menu
* Added system overclock support
* Added cores back to the image
* Added battery charger logic (use reboot or shutdown in simplemenu to change behavior)
* System should work with either 1 sdcard or 2 sdcard (SD1 koriki, SD2 roms)
  * By default roms use the batocera roms names and assume those are located in ``/userdata/roms/<system>``
  * You can use garlic or OFW images for the roms in the second SDCARD slot, but you will need to modify the paths in ``/userdata/system/.simplemenu/sections_groups``
* SDCard partition auto expand is current broken

2023/05/31
* ~~Added koriki_share_rg35xx.zip archive. This archive is intended to be used for a second sdcard. Uncompress at the top of the archive.~~
* It contains the cores, as well as the folder structure to add roms to the system

2023/05/30
* First development release
* Image automatically generated
* SDCARD expansion
* External SDCARD compatibility
* Only a few set of cores included

## Working:
* Navigation, rom selection, launching roms
* Brightness control
* Battery indicator
* Initial volume control via keys (menu does not reflect control change)
* Volume control for all emulators (thank to [adixal](https://codeberg.org/adixal) for this)
* System overclock control
* Retroarch menu can be opened with the ``Menu`` key. Quit with ``START``+``SELECT``
* Content auto-copied when a new SDCARD is inserted
* Auto mounts last partition of a second sdcard if present
* USB ethernet
* External controllers (in retroarch)

## Not working:
* ~~By default the system will start showing a message saying "No Games Found". Just press B to go back and select your rom. This is a bug that needs to be fixed in this version of simplemenu~~
* Some menus, settings are still incomplete
* Some emulators need to be updated
* ~~Standalone emulators require key mappings to be updated~~
* ~~Input event issues with retroarch (map retroarch configuration manually)~~
* Application launchers (e.g. Wifi, need to be updated)
* ~~When the console starts or the menu is opened after a game is closed simplemenu shows there are no games available. Press B to go back and then you can go into the selection.~~
* ~~Controls for standalone emulators are not properly mapped~~
* ~~Networking is still not configured~~
* ~~External controllers are not working at this point (kernel is in adb mode by default)~~
* ~~Selecting a core in simplemenu changes the background color (mostly cosmetic)~~

## Tweaking configuration:
* HOME is ``/userdata/system`` (also alias to /mnt/SDCARD)
* Simplemenu files are in ``$HOME/.simplemenu``
* Sections are in ``$HOME/.simplemenu/section_groups`` (self explanatory)
* Emulator launchers are in ``$HOME/.simplemenu/launchers`` (self explanatory, but look for the ones like neogeo_libretro that have been already adapted) 
* By default the firmware expects batocera-like roms partitions (with folders under roms like neogeo, mame, megadrive, etc.). It's possible add additional roms paths in ``$HOME/.simplemenu/section_groups`` .ini files. For instance, for master system you can see an example of batocera path (/userdata/roms/mastersystem) and Miyoo Koriki path (/userdata/Roms/MS). 
```[MASTER SYSTEM]
execs = /userdata/system/.simplemenu/launchers/picodrive_libretro,/userdata/system/.simplemenu/launchers/genesis_plus_gx_libretro,/userdata/system/.simplemenu/launchers/gearsystem_libretr
romDirs = /userdata/roms/mastersystem/,/userdata/Roms/MS
romExts = .zip,.sms,.7z
```
* The snapshots/art/thumbnail previews are by default under ``media/images`` (e.g. neogeo/media/images/mslug.png). You can change that path in $HOME/.simplemenu/config.ini and change the value of ``media_folder = media/images`` (for instance ``media_folder = Img`` is the default Koriki/miyoo path).

<hr /><em>This discussion was created from the release <a href='https://github.com/rg35xx-cfw/Koriki/releases/tag/koriki_rg35xx_v0.9'>Koriki for RG35XX Beta 0.8</a>.</em>

**Comment by fender357** [comment]
Should fix title from saying "0.8" to saying "0.9"

**Reply by acmeplus** [reply]
Thanks. Fixed it

**Comment by acmeplus** [comment]
Mupen64 config has changed, copy [this file](https://github.com/rg35xx-cfw/Koriki/releases/download/koriki_rg35xx_v0.9/mupen64plus.cfg) to  ``system/configs/mupen64/mupen64plus.cfg``

**Reply by SerJikBes** [reply]
same problem with the nintendo ds

**Comment by mrkllc** [comment]
Does anyone know how to input mapping keys in drastic? After updating to 0.9 it suddenly stopped working. Already tried copying old cfg file to SHARE/SYSTEM/CONFIGS/DRASTIC/CONFIG but still no luck. TIA.

**Reply by nhilez0017** [reply]
same problem..trying to remap it..i hope it will be successful..

**Reply by acmeplus** [reply]
Updated config for Drastic. Copy this [drastic.cfg](https://github.com/rg35xx-cfw/Koriki/releases/download/koriki_rg35xx_v0.9/drastic.cfg) to ``system/configs/drastic/config/drastic.cfg``

**Reply by mrkllc** [reply]
Thank you so much! I love your work.

**Comment by Maximus744** [comment]
On 0.9 from all available ports can run only spelunky, but dpad is not working there, and on other ports it's stuck on black screen or returns to simplemenu.

**Reply by acmeplus** [reply]
I’ll post updated launchers for those. I believe I had all of them working except hurrican.

**Reply by Maximus744** [reply]
and 
ppsspp sa needs fix control as well

**Reply by acmeplus** [reply]
Edit ``system/.simplemenu/launchers/ppsspp_sa`` and replace the line starting with SDL_VSYNC=1 with:
``SDL_VSYNC=1 SDL_RENDER_DRIVER=opengles2 PPSSPP "$1"``

The specific ppsspp configuration may still need to be updated, but that should get all buttons and dpad to respond.

**Reply by acmeplus** [reply]
For the ports, edit their bash launcher scripts, and replace the line starting with ``export SDL_GAMECONTROLLERCONFIG=...`` with: ``export SDL_GAMECONTROLLERCONFIG=$(grep "RG35XX" "${HOME}/.config/gamecontrollerdb.txt")``

This works for MalditaCastilla, CurseOfIssyos, and Spelunky. Hurrican does not work with that fix (buttons work but dpad does not respond). @JohnnyOnFlame mentioned Hurrican control mappings were a bit tricky...

**Reply by Maximus744** [reply]
Can confirm that psp controls now working fine, but ports controls still not

**Comment by slapjackios** [comment]
All of the new keybinds work perfectly, except for Menu+A for pausing the game. I tried it in several systems.
Also volume buttons still don't work after applying the bootfix which is a bummer.

**Reply by acmeplus** [reply]
The issue is that I assume you have a revision 6.0C in those boards menu, volume+/- and probably start and select have different mappings.

**Comment by NamVu69** [comment]
Copy uncompressed files from boot.taz.xz to sd card and it says not enough disk space.

**Reply by Level99OCR** [reply]
I also had this issue. I ended up doing a clean flash instead, unfortunately.

**Reply by acmeplus** [reply]
For v0.9 you want to upgrade without refreshing, you need to delete everything from the boot (batocera) partition and copy over the files extracted from boot.tar.xz. Then, you need to remove the system folder from the SHARE partition so it gets created again with the correct configuration files.

**Comment by dmauri82** [comment]
Where is the file bootfix_for_some_revisions.zip. Because in my case, the image not boot. Hungs in the initial black screen with all logos...
Tks very much!

**Reply by slapjackios** [reply]
I got it from the 0.8 release

**Reply by dmauri82** [reply]
tks! fix it! now the images boots but the volume controls doesn't respond (how the notes said).

**Comment by dmauri82** [comment]
where I can find a media pack with all the game preview images? (like mslug.png)

**Reply by slapjackios** [reply]
I found this github page that has loads of ready made images, but if the names don't match exactly, it would be a pain to rename all the files to match.

https://github.com/UpdogUpdogUpdog/miyoo-mini-imgs/releases/tag/stock_onionv2
